ocaml/xapi/xapi_pool.ml

```diff
@@ -2857,3 +2870,58 @@ let sync_updates ~__context ~self ~force =
         ~doc:"pool.sync_updates"
         ~op:`sync_updates
         (fun () -> create_pool_repository ~__context ~self:repository))
+
+let get_updates_handler (req : Http.Request.t) s _ =
+  debug "Pool.get_updates_handler: received request";
+  req.Http.Request.close <- true;
+  let query = req.Http.Request.query in
+  Xapi_http.with_context "Getting available updates" req s (fun __context ->
+      let all_hosts = Db.Host.get_all ~__context in
+      let hs =
+        match List.assoc "host_refs" query with
+          | v ->
+            List.map (fun ref_str -> Ref.of_string ref_str) (Astring.String.cuts ~sep:";" v)
+            |> (fun l ->
+                let is_invalid ref =
+                  not (Db.is_valid_ref __context ref && List.mem ref all_hosts)
+                in
+                match (List.exists is_invalid l), l with
+                | true, _ | false, [] -> None
+                | false, l -> Some l)
+          | exception Not_found -> Some all_hosts
+      in
+      match hs with
+      | Some hosts ->
+        let codes_of_404 =
```

Comment:
A better name would be simply `failures`. The name suggests currently that these errors are resulting from 404 errors but it is the reverse: these errors lead to a 404 response.

Reply:
What happens when `hosts = []`? Is this still working as intended?

Reply:
the case `hosts = []` is converted into a none just above here. it would probably be simpler to avoid the `list option` and just match using `[] -> ... | _::_ ->` instead